src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/crushmap/crushmap.component.ts

```diff
@@ -0,0 +1,69 @@
+import { Component, OnInit } from '@angular/core';
+
+import { NodeEvent, TreeModel } from 'ng2-tree';
+
+import { DashboardService } from '../../../shared/api/dashboard.service';
+
+@Component({
+  selector: 'cd-crushmap',
+  templateUrl: './crushmap.component.html',
+  styleUrls: ['./crushmap.component.scss']
+})
+export class CrushmapComponent implements OnInit {
+  panelTitle: string;
+  treeData: TreeModel;
+  crushMetaDatas: any[];
+  metadata: any;
+  metadataKeyMap: { [key: number]: number } = {};
+
+  constructor(private dashboardService: DashboardService) {
+    this.panelTitle = 'Cluster Crush Map';
+  }
+
+  ngOnInit() {
+    this.dashboardService.getHealth().subscribe((data: any) => {
+      this.crushMetaDatas = data.osd_map.tree.nodes || [];
+      this.treeData = this._abstractTreeData(data);
+    });
+  }
+
+  _abstractTreeData(data: any): TreeModel {
```

Comment:
Could you please limit the length of this function?

I've made a diff how you could enhance readability.

```diff
diff --git a/src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/crushmap/crushmap.component.ts b/src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/crushmap/crushmap.component.ts
index 05f33e5446..74e8edcb11 100644
--- a/src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/crushmap/crushmap.component.ts
+++ b/src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/crushmap/crushmap.component.ts
@@ -12,7 +12,6 @@ import { DashboardService } from '../../../shared/api/dashboard.service';
 export class CrushmapComponent implements OnInit {
   panelTitle: string;
   treeData: TreeModel;
-  crushMetaDatas: any[];
   metadata: any;
   metadataKeyMap: { [key: number]: number } = {};
 
@@ -22,48 +21,43 @@ export class CrushmapComponent implements OnInit {
 
   ngOnInit() {
     this.dashboardService.getHealth().subscribe((data: any) => {
-      this.crushMetaDatas = data.osd_map.tree.nodes || [];
       this.treeData = this._abstractTreeData(data);
     });
   }
 
   _abstractTreeData(data: any): TreeModel {
-    const cephNodes = data.osd_map.tree.nodes || [];
-    const nodesLength = cephNodes.length;
-    const tmpMap: { [key: number]: any } = {};
+    const nodes = data.osd_map.tree.nodes || [];
+    const rootNodeId = nodes[0].id;
+    const tree: { [key: number]: any } = {};
 
-    if (0 === nodesLength) {
-      return { value: 'No nodes!' };
+    if (nodes.length === 0) {
+      return {value: 'No nodes!'};
     }
+    nodes.reverse().forEach((node) => {
+      tree[node.id] = this.generateNodeLeaf(node, tree);
+    });
+    return tree[rootNodeId];
+  }
 
-    for (let i = nodesLength - 1; i > -1; i--) {
-      const id = cephNodes[i].id;
-      this.metadataKeyMap[id] = i;
-      let value: string = cephNodes[i].name + ' (' + cephNodes[i].type + ')';
-      const children: any[] = [];
-
-      if (cephNodes[i].crush_weight) {
-        value += '--' + cephNodes[i].crush_weight.toString();
-      }
+  private generateNodeLeaf(node, tree) {
+    const id = node.id;
+    this.metadataKeyMap[id] = node;
 
-      if (cephNodes[i].children) {
-        const len = cephNodes[i].children.length;
-        for (let j = 0; j < len; j++) {
-          children.push(tmpMap[cephNodes[i].children[j]]);
-        }
-        tmpMap[id] = { value, id, children };
-      } else {
-        tmpMap[id] = { value, id };
-      }
+    let value: string = node.name + ' (' + node.type + ')';
+    if (node.crush_weight) {
+      value += '--' + node.crush_weight.toString();
+    }
 
-      if (0 === i) {
-        return tmpMap[id];
-      }
+    const children: any[] = [];
+    if (node.children) {
+      node.children.forEach((childId) => {
+        children.push(tree[childId]);
+      });
     }
+    return {value, id, children};
   }
 
   onNodeSelected(e: NodeEvent) {
-    const index = this.metadataKeyMap[e.node.id];
-    this.metadata = this.crushMetaDatas[index];
+    this.metadata = this.metadataKeyMap[e.node.id];
   }
 }

```

Reply:
Thanks for your suggestion, I have modified it.